internal/rbd/group/volume_group.go

```diff
@@ -82,6 +85,12 @@ func GetVolumeGroup(
 		return nil, fmt.Errorf("failed to get volume attributes for id %q: %w", vg, err)
 	}
 
+	if _, err := vg.GetName(ctx); err != nil {
```

Comment:
This doesnt seems to be the right fix. i see that `getVolumeGroupAttributes` we are comparing with `librbd.ErrNotFound` i dont think it ever returns librbd error message, either it returns the pooNotfound or the key not found error message, which need to be returned and the caller need to check for it like we are doing in other delete volume or delete snapshot RPC call

Reply:
I don't think it ever returns either of the 2 errors as well. We just log in case either of these 2 errors are found, and continue to update the attributes struct.
Can you please link the other delete RPC call where we are actually doing it that way?

Reply:
So, we return PoolNotFound error if we fail to get the pool is not found as present in the param.
But, I don't see that we are actually returning the `ErrKeyNotFound` error, apart from the `FetchAttribute` function that is used in the nfs controller.
So, not sure if I am missing something in the design aspect or is this actually a bug?

Reply:
Yes thats correct, i missed it but there are couple of things still bothering me

1. The `errors.Is(err, librbd.ErrNotFound) ` check above is never going to get hit.
2. The `GetName` returns error message as bug which is not the case as its expected.
3. https://github.com/ceph/ceph-csi/blob/a32ba13045f4f433d4c00f568c67209556428a76/internal/csi-addons/rbd/volumegroup.go#L193-L200 this is wrong as we cannot assume the error as not found

Reply:
1. I agree, even I don't see that erorr being returned anywhere. What error shall we check for instead?
2. Becuase, it is used while CreateVolumeGroup RPC call to veirfy that the name was set correctly when the VolumeGroup is being created. But, as it's expected bug we are handling it in the parent function. How else do you think we should handle this case?
3. Agreed, that's why I have updated that code to return the error as `InternalError` above.

Reply:
lets hear from @nixpanic

Reply:
`librbd.ErrNotFound` can be returned by go-ceph operations. In recent go-ceph, all `ErrNotFound` (in the rbd, cephfs and rados packages) map to `-C.ENOENT`. These errors can be mixed, and still can be compared to eachother as long as the cause (not found errno) is the same.

Ideally all errors that are returned include the underlying error with `%w` formatting. That way, a single error can be of multiple types, and comparing them is less prone to problems.

Some go-ceph functions are composed out of multiple C calls to librados and/or to librbd/libcephfs. It is not trivial to know which error is returned when, just "not found" is sufficient and should be treated the same way for librados and librbd.

A similar problem surfaces sometimes in Ceph-CSI. Some operations return the received error directly, others replace it with their own variant. This makes it a good source of bugs, as checking for the error type is non-trivial.

The safest would be to have dedicated errors per package, and only return those (maybe only for exported functions?). This can be done by starting to use the `%w` formatting, and even including matching errors with the `%w` formatting in the exported error variables.

```go
var (
	ErrRBDGroupNotConnected = fmt.Errorf("%w: RBD group is not connected", librados.ErrNotConnected)
	ErrRBDGroupNotFound     = fmt.Errorf("%w: RBD group not found", librbd.ErrNotFound)
)
```

This way, `err = ErrRBDGroupNotFound; errors.Is(err, librbd.ErrNotFound)` will still be true, and changes to the rest of the code can be done piece by piece. Eventually only `errors.Is(err, group.ErrRBDGroupNotFound)` would then be used outside of the `internal/rbd/group` package.

Reply:
I am wondering if having a name check that return the `librbd.ErrNotFound` error from the `getVolumeAttributes` function would be a better choice? In that way, we can just compare the `ErrRBDGroupNotFound ` error from that function and return?

Reply:
I have updated the code to reflect my above statement, please check again once.